src/hivpy/resistance_mutations.py

```diff
@@ -0,0 +1,388 @@
+        pop.init_variable(col.RT103_MUTATION, False)
+        pop.init_variable(col.RT181_MUTATION, False)
+        pop.init_variable(col.RT190_MUTATION, False)
+        pop.init_variable(col.PR32_MUTATION, False)
+        pop.init_variable(col.PR33_MUTATION, False)
+        pop.init_variable(col.PR46_MUTATION, False)
+        pop.init_variable(col.PR47_MUTATION, False)
+        pop.init_variable(col.PR50L_MUTATION, False)
+        pop.init_variable(col.PR50V_MUTATION, False)
+        pop.init_variable(col.PR54_MUTATION, False)
+        pop.init_variable(col.PR76_MUTATION, False)
+        pop.init_variable(col.PR82_MUTATION, False)
+        pop.init_variable(col.PR84_MUTATION, False)
+        pop.init_variable(col.PR88_MUTATION, False)
+        pop.init_variable(col.PR90_MUTATION, False)
+        pop.init_variable(col.IN118_MUTATION, False)
+        pop.init_variable(col.IN140_MUTATION, False)
+        pop.init_variable(col.IN148_MUTATION, False)
+        pop.init_variable(col.IN155_MUTATION, False)
+        pop.init_variable(col.IN263_MUTATION, False)
+
+    def get_matrix_val(self, matrix, active_drugs, cont_on_art, adherence, adherence_tm1,
+                       on_nev=None, on_efa=None):
+        """
+        Returns a value from either the new mutation matrix or the CD4 delta matrix given the input parameters.
+        """
+        # find matrix indices
+        active_drug_index = np.digitize(active_drugs, self.active_drug_bins)
+        cont_on_art_index = np.digitize(cont_on_art.years(), self.cont_on_art_bins)
+        adherence_index = np.digitize(adherence, self.adherence_bins)
+
+        # adjust adherence index if taking specific ART drugs (only relevant to new mutation probability)
+        if adherence_index == 0 and (on_nev or on_efa):
+            adherence_index += 1
+
+        # lookup matrix value
+        x = matrix[active_drug_index][cont_on_art_index][adherence_index]
+        # account for adherence last time step when on ART for 3-6 months
+        if cont_on_art_index == 1:
+            adherence_tm1_index = np.digitize(adherence_tm1, self.adherence_bins)
+            x = x[adherence_tm1_index]
+
+        return x
+
+    def viral_load(self, pop: Population, sub_pop):
+        """
+        Update viral load for HIV+ individuals.
+        """
+        # get viral load outcomes
+        viral_load = pop.col_apply([col.NUM_ACTIVE_DRUGS, col.CONT_ON_ART,
+                                    pop.get_correct_column(col.ART_ADHERENCE, dt=0),
+                                    pop.get_correct_column(col.ART_ADHERENCE, dt=1),
+                                    col.MAX_VIRAL_LOAD],
+                                   self.calc_viral_load, sub_pop=sub_pop)
+
+        pop.set_present_variable(col.VIRAL_LOAD, viral_load, sub_pop)
+
+    def get_viral_load_matrix(self, max_viral_load):
+        # viral_load_matrix[active_drugs][cont_on_art][adherence]
+        # FIXME: is there a better way to do this?
```

Comment:
This is ridiculously inefficient for population sizes of 100k+, but even switching to a static matrix containing a, b, and c values for a `a * max_viral_load + b + c * min_vl_on_art` calculation isn't much better because the main problem appears to be that a separate calculation is happening for every person in the population.

I think it looks like the vl calculation will have to be separated into more `transform_group`-able chunks to improve performance.

Reply:
Using a static matrix and pre-calculating all matrix indices reduces the runtime of one iteration of viral load calculations on my machine by approximately 60% (46s > 18s with a sub-population size of 1 000 000).